smartlearner/interfaces/loss.py

```diff
@@ -5,20 +6,29 @@ class Loss(object):
     def __init__(self, model, dataset):
         self.model = model
         self.dataset = dataset
-        self.target = dataset.symb_targets
         self.consider_constant = []  # Part of the computational graph to be considered as a constant.
+        self.graph_updates = OrderedDict()
+        self._gradients = None
+
+    @property
+    def target(self):
+        return self.dataset.symb_targets
 
     def get_graph_output(self):
         output, updates = self.model.get_model_output(self.dataset.symb_inputs)
         return self._loss_function(output), updates
 
-    def get_gradients(self):
-        cost, updates = self.get_graph_output()
-        gparams = T.grad(cost=cost,
-                         wrt=self.model.parameters,
-                         consider_constant=self.consider_constant)
-        gradients = dict(zip(self.model.parameters, gparams))
-        return gradients, updates
+    @property
+    def gradients(self):
+        if self._gradients is None:
+            cost, updates_from_model_output = self.get_graph_output()
+            gparams = T.grad(cost=cost,
+                             wrt=self.model.parameters,
+                             consider_constant=self.consider_constant)
+            self._gradients = dict(zip(self.model.parameters, gparams))
+            self.graph_updates.update(updates_from_model_output)
+
+        return self._gradients
```

Comment:
Forcing to return the gradients and the updates was the problem in the first place. I wanted to print the norm of the gradients so I had to manually call the `get_gradients` function and had to manually add the updates returned by the function to the trainer updates before compiling the graph!

I think that if every object has an `updates` variable (above I said property but I meant variable), which is an `OrderedDict`, then it is the responsibility of the object to fill its own `updates` dictionary. After that, when it's time to compile the graph, the trainer (or whatever the object responsible of creating the graph) will fetch every updates using the chain of command principle (e.g. asks the optimizer which asks the loss which asks the model and so on).

Of course, the user can forget to populate the `updates` variable of its object.